### DISCUSSION TITLE
Specify different versions of dependencies for different versions

### DISCUSSION THREAD

**Discussion by brenthuisman** [spack/spack#26848] [Q&A] [answered]
Dependencies of a package evolve over time, e.g. require more recent versions. However, I might not want to change the dependency versions for older versions of the package, e.g. if you build `package@v0.5` it should pull in `dep@v2.6.2`, and `package@v0.6` should pull `dep@2.9:`. I've gone through the [dep section in the packaging guide](https://spack.readthedocs.io/en/latest/packaging_guide.html#dependencies) but I can't find a way to do this. It seems I can only specify dependencies at the package level (at the variant level), but not per version. Is my understanding correct, or am I overlooking something?

**Comment by haampie** [comment, marked as answer]
Do you require an upper bound or exact version of a dependency, or do you prefer to rebuild fewer specs?

For the former: try something like this: `depends_on('dep@:2.6', when='@:0.5')`, it's typically better to work with ranges than exact versions, and here you just set an upper bound so that any `package@0.5.x` works with any `dep@2.6.y`.

For the latter: this is a matter of concretization, and is being worked on, something like `spack concretize --reuse` to add a concretization criterion which favors reuse over newer versions. Until then, just be explicit on the command line, or use environments if you are tired of long specs on the command line.

Please don't add strict `depends_on` rules to minimize builds, cause it would the package miss out on bugfix releases.

**Reply by brenthuisman** [reply]
Thanks, I didn't catch the `when` flag! `when` does not let me exclude anything, right? For instance: exclude non-numeric versions?

Reasons are two-fold:
1. As the Spack ecosystem ages, dep requirements of older versions is going to significantly diverge from current dep version requirements. We don't test older versions with current dependencies as a matter of course, and I'm not sure we are going to. Freezing older versions at specific dependency versions may be the only practical way we can insure those older versions remain usable/behave predictably/have been actually tested. In addition: we're not going to be in control of all build envs. I'm not fond of relying on users/builders speccing the env out in excruciating detail, because I'm not sure it'll be done correctly (it's hard for me!); freezing the deps of older versions seems like a better idea to ensure `spack install my_package@some_old_version` results in something stable over time. Concrete use-case: processing pipelines that researchers store and re-use perhaps for decades. I'm not sure what's optimal here yet either, so any opinions on how to keep such pipelines working over longer timespans is appreciated!
2. Many packages define a `master` version, which will get you their current `master` branch. Together with the fact that Spack file updates trail project releases by some time, this may mean in practice that specific versions may be pulling in unspecific dependency versions. That I am not fond of. When building our package's master branch I'm OK with pulling in deps master branches, but not for any version release. The only way to prevent this right now seems to add an upper limit to `when`, correct?

**Reply by haampie** [reply]
> Concrete use-case: processing pipelines that researchers store and re-use perhaps for decades

In that case use environments, concretize once and for all, store the `lock` file in version control. That is, don't reconcretize if you want to work with the same environment.

> specific versions may be pulling in unspecific dependency versions

There is work in progress about better support for git commit sha's as versions: https://github.com/spack/spack/pull/25981
